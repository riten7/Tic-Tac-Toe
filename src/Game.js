import React from 'react';
import Board from './Board';
import { calculateWinner } from './utility';

const initialState = {
  squares: Array(16).fill(null),
  IsXNext: true,
};

const Game = () => {
  const [gameState, setGameState] = React.useState(initialState);

  const { winner, winnerRow } = calculateWinner(gameState.squares);

  const gameStatus = () => {
    let status = "";
    if (winner) {
      status = ` Congratulation! ${winner} have won`;
    } else {
      status = `Next player: ${gameState.IsXNext ? 'X' : 'O'}`;
    }
    return status;
  }

  const reset = () => {
    setGameState(initialState);
  }

  const handleClick = (index) => {
    const squares = gameState.squares.slice();
    const winner = calculateWinner(squares).winner;

    if (winner || squares[index]) return;
    squares[index] = gameState.IsXNext ? 'X' : 'O';
    setGameState({
      squares: [...squares],
      IsXNext: !gameState.IsXNext,
    })
  }

    return (
      <div className="game">
        <div className="game-status">{gameStatus()}</div>
        <div className="game-board">
          <Board
            squares={gameState.squares}
            winnerSquares={winnerRow}
            onClick={index => handleClick(index)}
          />
        </div>
        <div className="game-info">
          <button className="button button--new-game" onClick={() => reset()}>
            Reset
          </button>
        </div>
      </div>
    );
  };

  export default Game;
