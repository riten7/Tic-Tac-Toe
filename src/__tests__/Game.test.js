import React from "react";
import { mount } from "enzyme";
import Game from "../Game";

describe("Component renders properly", () => {
  const wrapper = mount(<Game />);
  it("Game renders successfully", () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('.game-status').text()).toBe('Next player: X');
  });

  it("click on square board runs well", () => {
    wrapper.find('.square').at(0).simulate('click');
    expect(wrapper.find('.game-status').text()).toBe('Next player: O');
  });

  it("click on start game reset board", () => {
    wrapper.find('.button--new-game').simulate('click');
    expect(wrapper.find('.game-status').text()).toBe('Next player: X');
    expect(wrapper.find('.square').at(0).text()).toBe('');
  });

  it("test if the winner is shown when one player won", () => {
    wrapper.find('.square').at(3).simulate('click');
    wrapper.find('.square').at(5).simulate('click');
    wrapper.find('.square').at(7).simulate('click');
    wrapper.find('.square').at(4).simulate('click');
    wrapper.find('.square').at(11).simulate('click');
    wrapper.find('.square').at(8).simulate('click');
    wrapper.find('.square').at(15).simulate('click');
    wrapper.find('.square').at(12).simulate('click');
   
    expect(wrapper.find('.game-status').text()).toBe(' Congratulation! X have won');
  });
});