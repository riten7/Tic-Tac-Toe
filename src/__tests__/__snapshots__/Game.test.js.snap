// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders properly Game renders successfully 1`] = `"<div class=\\"game\\"><div class=\\"game-status\\">Next player: X</div><div class=\\"game-board\\"><div class=\\"board\\"><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button><button class=\\" square\\"></button></div></div><div class=\\"game-info\\"><button class=\\"button button--new-game\\">Reset</button></div></div>"`;
